src/libUtils/UpgradeManager.cpp

```diff
@@ -15,148 +15,370 @@
-    curOffset += sizeof(uint32_t);
-    SetNumber<uint32_t>(dst, curOffset, m_minor, sizeof(uint32_t));
-    curOffset += sizeof(uint32_t);
-    SetNumber<uint32_t>(dst, curOffset, m_fix, sizeof(uint32_t));
-    curOffset += sizeof(uint32_t);
-    SetNumber<uint64_t>(dst, curOffset, m_upgradeDS, sizeof(uint64_t));
-    curOffset += sizeof(uint64_t);
-    SetNumber<uint32_t>(dst, curOffset, m_commit, sizeof(uint32_t));
-    curOffset += sizeof(uint32_t);
+        if (2 == find)
+        {
+            find = 0;
+            downloadFilePaths.emplace_back(*tok_iter);
+            continue;
+        }
 
-    return SIZE;
-}
+        cur = *tok_iter;
 
-/// Implements the Deserialize function inherited from Serializable.
-int SWInfo::Deserialize(const std::vector<unsigned char>& src,
-                        unsigned int offset)
-{
-    LOG_MARKER();
+        if (cur == "browser_download_url")
+        {
+            find = 1;
+            continue;
+        }
+    }
 
-    unsigned int curOffset = offset;
+    string downloadFilePath;
 
-    try
+    for (auto s : downloadFilePaths)
     {
-        m_major = GetNumber<uint32_t>(src, curOffset, sizeof(uint32_t));
-        curOffset += sizeof(uint32_t);
-        m_minor = GetNumber<uint32_t>(src, curOffset, sizeof(uint32_t));
-        curOffset += sizeof(uint32_t);
-        m_fix = GetNumber<uint32_t>(src, curOffset, sizeof(uint32_t));
-        curOffset += sizeof(uint32_t);
-        m_upgradeDS = GetNumber<uint64_t>(src, curOffset, sizeof(uint64_t));
-        curOffset += sizeof(uint64_t);
-        m_commit = GetNumber<uint32_t>(src, curOffset, sizeof(uint32_t));
-        curOffset += sizeof(uint32_t);
+        if (string::npos != s.rfind(fileTail))
+        {
+            downloadFilePath = s;
+            break;
+        }
     }
-    catch (const std::exception& e)
+
+    string fileName = downloadFilePath.substr(downloadFilePath.rfind('/') + 1);
+
+    /// Get the redirection url (if applicable)
+    curl_easy_setopt(curl, CURLOPT_URL, downloadFilePath.data());
+    res = curl_easy_perform(curl);
+
+    if (res != CURLE_OK)
     {
         LOG_GENERAL(WARNING,
-                    "Error with SWInfo::Deserialize." << ' ' << e.what());
-        return -1;
+                    "curl_easy_perform() failed: " << curl_easy_strerror(res));
+        curl_easy_cleanup(curl);
+        curl_global_cleanup();
+        return nullptr;
     }
 
-    return 0;
-}
+    long response_code;
+    res = curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response_code);
 
-/// Less-than comparison operator.
-bool SWInfo::operator<(const SWInfo& r) const
-{
-    return tie(m_major, m_minor, m_fix, m_upgradeDS, m_commit)
-        < tie(r.m_major, r.m_minor, r.m_fix, r.m_upgradeDS, r.m_commit);
-}
+    if ((res != CURLE_OK) || ((response_code / 100) == 3))
+    {
+        char* location;
+        curl_easy_getinfo(curl, CURLINFO_REDIRECT_URL, &location);
```

Comment:
Error checking is missing. This would cause memory overflow when the uninitialized variable `location` is assigned to `downloadFilePath` and `downloadFilePath` is used aftewards.

https://curl.haxx.se/libcurl/c/curl_easy_getinfo.html
> The data pointed-to will be filled in accordingly and can be relied upon only if the function returns CURLE_OK.